source/robot_model/src/Model.cpp

```diff
@@ -260,9 +263,114 @@ std::vector<state_representation::CartesianPose> Model::forward_geometry(const s
+    }
+  }
+  return W_b;
+}
+
+Eigen::VectorXd Model::cwln_repulsive_potential_field(const state_representation::JointPositions& joint_positions, double margin) {
+  Eigen::VectorXd Psi(this->robot_model_.nq);
+  Eigen::VectorXd q = joint_positions.get_positions();
+  for (int i = 0; i < this->robot_model_.nq; i++) {
+    Psi[i] = 0;
+    if(q[i] < this->robot_model_.lowerPositionLimit[i] + margin) {
+      Psi[i] = this->robot_model_.upperPositionLimit[i] - margin - std::max(q[i], this->robot_model_.lowerPositionLimit[i]);
+    } else if(this->robot_model_.upperPositionLimit[i] - margin < q[i]) {
+      Psi[i] = this->robot_model_.lowerPositionLimit[i] + margin - std::min(q[i], this->robot_model_.upperPositionLimit[i]);
+    }     
+  }
+  
+  return Psi;
+}
+
+state_representation::JointPositions Model::inverse_geometry(const state_representation::CartesianState& desired_cartesian_state,
+                                                              const state_representation::JointState& current_joint_state,
+                                                              std::string frame_name,
+                                                              const Model::InverseGeometryParameters& params) {
+
+  if (frame_name.empty()) {
+    // get last frame if none specified
+    frame_name = this->robot_model_.frames.back().name;
+  } else {
+    // throw error if specified frame does not exist
+    if (!this->robot_model_.existFrame(frame_name)) {
+      throw (exceptions::FrameNotFoundException(frame_name));
+    }
+  }
+
+  const int frame_id = this->robot_model_.getFrameId(frame_name);
+  
+  // 1 second for the Newton-Raphson method
+  const std::chrono::nanoseconds dt((int)(1e9));
+
+  state_representation::JointPositions q(current_joint_state);
+  state_representation::JointVelocities dq(this->get_robot_name(), this->get_number_of_joints());
+  state_representation::Jacobian J(this->get_robot_name(), this->robot_model_.nq, frame_name);
+  Eigen::MatrixXd J_b = J.data();
+  Eigen::MatrixXd JJt(6,6);
+  Eigen::MatrixXd W_b = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::MatrixXd W_c = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq);
+  Eigen::VectorXd psi(this->robot_model_.nq);
+  Eigen::VectorXd err(6);
+  state_representation::CartesianPose X(this->get_robot_name());
+  const state_representation::CartesianPose X_d(desired_cartesian_state);
+  
+  int i = 0;
+  while (i < params.max_number_of_iterations) {
+    X = this->forward_geometry(q, frame_id);
+    
+    err = ((X_d - X)/dt).data();
+
+    if(err.array().abs().maxCoeff() < params.tolerance) {
+      return q;
+    }
+
+    J = this->compute_jacobian(q);
+    
+    W_b = this->cwln_weighted_matrix(q, params.margin);
+    W_c = Eigen::MatrixXd::Identity(this->robot_model_.nq, this->robot_model_.nq) - W_b;
+    psi = params.gamma * this->cwln_repulsive_potential_field(q, params.margin);
+    J_b = J.data() * W_b;
```

Comment:
@buschbapti is this `.data()` still necessary? One of your operators does this directly no?

Reply:
it is not indeed I was going to comment that

Reply:
Here `J_b` is of type `MatrixXd` and not `Jacobian` for that I put `.data()`.
I don't use `Jacobian` because if I want to evaluate the pseudoinverse and add some damping to `J*Jt` I need to use `MatrixXd` and for that `err` and `psi` are not `CartesianTwist` but `VectorXd`.

Reply:
It's all good how you do it but Baptiste implemented the `*` operator for the jacobian, such that you can directly multiply it with an Eigen matrix. The `data()` in line 352 and 356 are therefore unnecessary

Reply:
no but PR #92 merged yesterday allows you to do `J * W_b` and that returns a `MatrixXd` so you are all good

Reply:
than it returns a `Jacobian`or a `MatrixXd`, because I want a `MatrixXd`

Reply:
Trust us, I even tested it on my computer. You can remove the two `data()`. And here is the implementation of the operation: https://github.com/epfl-lasa/control_libraries/blob/8433bacc53f5285265c29b288abf7b5bf92f7d9b/source/state_representation/src/robot/Jacobian.cpp#L153

Reply:
ah okok I trust you.